### PR TITLE
Dltp 767 gather curate holding informati

### PR DESCRIPTION
**Pull request by banurekha**
Gathering all holding information as per report

### CODE REVIEW

File: .hound.yml [hunk 1]
```diff
@@ -63,7 +65,7 @@ Metrics/LineLength:
 
 Metrics/AbcSize:
   Enabled: true
-  Max: 16
```
Comment (jeremyf) [inline .hound.yml]: What is the threshold number. I thought it was 16.x. Round the reported number up to the next digit. In doing this, does the `app/services/fedora_object_harvester.rb` now pass?

File: .started-issues [hunk 2]
```diff
@@ -4,4 +4,4 @@
 35
 37
 39
-DLTP-766
\ No newline at end of file
+DLTP-766DLTP-767
```
Comment (jeremyf) [inline .started-issues]: Something odd is going on here. I think you need a new line.

File: app/models/fedora_object.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 # A reporting record for an object persisted in Fedora.
 class FedoraObject < ActiveRecord::Base
   has_many :fedora_object_aggregation_keys, dependent: :destroy
+
+  def self.generic_files(options = {})
+    where(af_model: 'GenericFile').where('obj_modified_date <= :as_of', as_of: options.fetch(:as_of)).group_by(&options.fetch(:group).to_sym)
```
Comment (jeremyf) [inline app/models/fedora_object.rb]: I find it more difficult to read inline `options.fetch` calls. Create variables local variables for as_of and group_by. Also, in separating the fetch calls, if an exception is thrown on KeyError, you will know which one and see the immediate problem.

```ruby
def self.generic_files(options={})
  as_of = options.fetch(:as_of)
  group_by = options.fetch(:group_by).to_sym
  where(af_model: 'GenericFile').where('obj_modified_date <= :as_of', as_of: as_of).group_by(&group_by)
end
```

File: app/models/fedora_object_aggregation_key.rb [hunk 4]
```diff
@@ -1,4 +1,13 @@
 # Value object for storing storing Fedora identifiers
 class FedoraObjectAggregationKey < ActiveRecord::Base
   belongs_to :fedora_object
+
+  def self.aggregation_by(options = {})
+    sql = "select b.aggregation_key as aggregation_key, count(b.fedora_object_id) as object_count from #{quoted_table_name} b
```
Comment (jeremyf) [inline app/models/fedora_object_aggregation_key.rb]: Create local variables based on the options that are fetched. (See above comment for reasoning).

File: app/services/metrics_report.rb [hunk 5]
```diff
@@ -1,19 +1,31 @@
 require 'erb'
+require 'action_view'
 
 # Create Metrics Report for given dates
 class MetricsReport
   # Helper class to store report details
   class MetricsDetail
     attr_reader :report_start_date, :report_end_date, :fedora_count,
-                :fedora_size, :storage
+                :fedora_size, :storage, :gf_by_holding_type
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: What does `gf_by_hold_type` mean? Favor an explicit variable name instead of shorthand.

File: app/services/metrics_report.rb [hunk 6]
```diff
@@ -30,14 +42,37 @@ def generate_report
...
     rescue StandardError => e
       @exceptions << e.inspect
     end
     report_any_exceptions
   end
 
-  private
+  # get all the count for administrative unit and present them hierarchical order
+  # with department wide count and total count
+  def administrative_unit_presenter(unit = metrics.obj_by_administrative_unit, html = "")
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: The method name presenter seems weird. Presenters are often classes that have information for rendering. I believe you are asking for `report_administrative_unit_as_html`

File: app/services/metrics_report.rb [hunk 7]
```diff
@@ -55,6 +90,72 @@ def storage_information_for(storage_type)
...
 
+  def holding_object_counts
+    metrics.items_added_count = FedoraObject.where('obj_ingest_date <= :end_date AND  obj_ingest_date >= :start_date',
+                                                   start_date: metrics.report_start_date, end_date: metrics.report_end_date).count
+
+    metrics.items_modified_count = FedoraObject.where('obj_modified_date <= :end_date AND  obj_modified_date >= :start_date',
+                                                      start_date: metrics.report_start_date, end_date: metrics.report_end_date).count
+  end
+
+  def gf_info_for(holding_type)
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: Favor explicit name instead of short variable name. In 5 months when we go back to read this, what will gf mean?

File: app/services/metrics_report.rb [hunk 8]
```diff
@@ -55,6 +90,72 @@ def storage_information_for(storage_type)
...
+    gf_by_type = FedoraObject.generic_files(as_of: metrics.report_end_date, group: holding_type)
+    holding_type_objects = {}
+    gf_by_type.each do |type, objects|
+      holding_type_objects[type] = ReportingStorageDetail.new(count: objects.count, size: objects.map(&:bytes).sum)
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: I'm okay with leaving this, but `objects.map(&:bytes).sum` will loop through objects twice.

File: app/services/metrics_report.rb [hunk 9]
```diff
@@ -55,6 +90,72 @@ def storage_information_for(storage_type)
...
+
+  # Get administrative count as a recursive hierarchical hash
+  def administrative_unit_count
+    administrative_unit_array = []
+    administrative_units = FedoraObjectAggregationKey.aggregation_by(as_of: metrics.report_end_date,
+                                                                     predicate: 'creator#administrative_unit')
+    administrative_units.each do |administrative_unit|
+      aggregation_keys = administrative_unit.fetch('aggregation_key').split('::')
+      # remove header (University of NotreDame)
+      aggregation_keys.shift
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: It is confusing having the aggregation_keys shift happen in two places. Here and in `convert_adminstrative_unit_to_hash`. Please move this to that method.

File: app/templates/periodic_metrics.html.erb [hunk 10]
```diff
@@ -19,16 +17,152 @@
...
+        <th>Size (Bytes)</th>
+        <th>Size (GB)</th>
+      </tr>
+      </thead>
+      <tbody>
+      <% total_count = 0 -%>
+      <% total_size = 0 -%>
+      <% holding_type.each do |mime_type, generic_files| -%>
+        <tr>
+          <% total_count = total_count + generic_files.count -%>
```
Comment (jeremyf) [inline app/templates/periodic_metrics.html.erb]: `total_count += generic_files.count` is shorthand for the above

File: app/templates/periodic_metrics.html.erb [hunk 11]
```diff
@@ -19,16 +17,152 @@
...
+    <tbody>
+      <% metrics.obj_by_academic_status.each do |status| -%>
+        <tr>
+          <td> <%= status.fetch("aggregation_key") %> </td>
+          <td> <%= status.fetch("object_count") %> </td>
+        </tr>
+      <% end -%>
+      <tr>
+        <td> Total </td>
+        <td> <%= metrics.obj_by_academic_status.map{|t|t.fetch("object_count")}.sum %> </td>
```
Comment (jeremyf) [inline app/templates/periodic_metrics.html.erb]: Favor a method on metrics; Also consider that you are iterating twice through this array.

File: db/schema.rb [hunk 12]
```diff
@@ -14,72 +14,71 @@
 ActiveRecord::Schema.define(version: 20170112165749) do
 
   create_table "curate_storage_details", force: :cascade do |t|
-    t.string   "storage_type", limit: 255, null: false
-    t.integer  "object_count", limit: 4,   null: false
-    t.integer  "object_bytes", limit: 8,   null: false
-    t.date     "harvest_date",             null: false
-    t.datetime "created_at",               null: false
-    t.datetime "updated_at",               null: false
+    t.string   "storage_type",           null: false
```
Comment (jeremyf) [inline db/schema.rb]: Was this generated via `rake db:migrate`

File: db/migrate/20170112165749_add_predicate_name_to_fedora_object_aggregation_key.rb [hunk 13]
```diff
@@ -1,6 +1,6 @@
 class AddPredicateNameToFedoraObjectAggregationKey < ActiveRecord::Migration
   def change
-    add_column :fedora_object_aggregation_keys, :predicate_name, :string, null: false
+    add_column :fedora_object_aggregation_keys, :predicate_name, :string, default: ''
```
Comment (jeremyf) [inline db/migrate/20170112165749_add_predicate_name_to_fedora_object_aggregation_key.rb]: Why is this change here? Why the shift from `null: false` to `default: ''`?

Since this migration has already run in production, the change would not be re-run. If this is needed, you'll need to create another migration.

File: app/templates/periodic_metrics.html.erb [hunk 14]
```diff
@@ -19,16 +17,152 @@
...
+        <th>Size (GB)</th>
+      </tr>
+      </thead>
+      <tbody>
+      <% total_count = 0 -%>
+      <% total_size = 0 -%>
+      <% holding_type.each do |mime_type, generic_files| -%>
+        <tr>
+          <% total_count = total_count + generic_files.count -%>
+          <% total_size = total_size +  generic_files.size -%>
```
Comment (jeremyf) [inline app/templates/periodic_metrics.html.erb]: `total_size += generic_files.size`

File: app/templates/periodic_metrics.html.erb [hunk 15]
```diff
@@ -19,16 +17,152 @@
     </tr>
     </thead>
     <tbody>
+      <% total_count = 0 %>
+      <% total_size = 0 %>
       <% metrics.storage.each do |storage_type, storage_details| -%>
         <tr>
+          <% total_count = total_count + storage_details.count -%>
```
Comment (jeremyf) [inline app/templates/periodic_metrics.html.erb]: `total_count += storage_details.count` is shorthand for the above

File: app/templates/periodic_metrics.html.erb [hunk 16]
```diff
@@ -19,16 +17,152 @@
...
+          <% total_size = total_size +  storage_details.size -%>
```
Comment (jeremyf) [inline app/templates/periodic_metrics.html.erb]: `total_size += storage_details.size` is shorthand for the above

File: app/services/metrics_report.rb [hunk 17]
```diff
@@ -55,6 +90,72 @@ def storage_information_for(storage_type)
...
+      administrative_unit_array << convert_administrative_unit_to_hash(aggregation_keys, administrative_unit.fetch('object_count'))
+    end
+    metrics.obj_by_administrative_unit = administrative_unit_array.reduce(&:deep_merge)
+  end
+
+  def convert_administrative_unit_to_hash(administrative_units, count)
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: It appears that we are assuming a depth of 3 or less. Is there consideration for 4 (as those won't show up in our reporting)?

File: app/services/metrics_report.rb [hunk 18]
```diff
@@ -30,14 +42,37 @@ def generate_report
...
-  private
+  # get all the count for administrative unit and present them hierarchical order
+  # with department wide count and total count
+  def administrative_unit_presenter(unit = metrics.obj_by_administrative_unit, html = "")
+    unit.each do |unit_name, count_by_administrative_unit|
+      # if given administrative_unit_hash is department hash, add department name and total count
+      if count_by_administrative_unit.is_a?(Hash)
+        count_by_department = count_by_administrative_unit.values
+        metrics.administrative_units_count = metrics.administrative_units_count + count_by_department.sum
+        html << "<tr class=department> \n <td> #{unit_name} </td> \n <td> #{count_by_department.sum} </td>\n </tr> \n"
```
Comment (jeremyf) [inline app/services/metrics_report.rb]: It looks like the html is the same as below but includes the `class=department`; Can you encode this into a method? I am a bit confused as the type is deparment but the variable is `unit_name`. How can this be clarified.

File: db/schema.rb (hunk 12 above)
Comment (banurekha) [inline db/schema.rb]: Yes.. There some weird things going on when I merged it from 819 bendo_type changes so had to make this

File: db/migrate/20170112165749_add_predicate_name_to_fedora_object_aggregation_key.rb (hunk 13 above)
Comment (banurekha) [inline db/migrate/20170112165749_add_predicate_name_to_fedora_object_aggregation_key.rb]: This is not in production yet and I changed it as part of DLTP-766.. There some weird things going on when I merged it from 819 bendo_type changes so had to make this again

File: app/services/metrics_report.rb (hunk 17 above)
Comment (banurekha) [inline app/services/metrics_report.rb]: Yes.. currently we support only three level in administrative unit. University::College::Department. So I went by that assumption

File: app/services/metrics_report.rb (hunk 18 above)
Comment (banurekha) [inline app/services/metrics_report.rb]: There has to be some styling has to happen here to separate each college:department list.. one total is sum of all department and other one is by department count, I can add this comment to explain
